### ISSUE TITLE
Nightly build heisenbugs

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2746]
These don't happen very often, but should be investigated:

on macos/x86.64:

```
==== resource:basis/io/monitors/monitors-tests.factor

resource:basis/io/monitors/monitors-tests.factor: 140

Unit Test: {
    { t }
    [
        [
            [
                <promise> "p" set "." t <monitor> "m" set
                10 seconds "m" get set-timeout [
                    [ "m" get next-change ] [ ] recover
                    "p" get fulfill
                ] in-thread 250 milliseconds sleep
                "m" get dispose
                "p" get 10 seconds ?promise-timeout
                already-disposed?
            ] with-monitors
        ] with-test-directory
    ]
}

=== Expected:
t
=== Got:
f
(U) [ c-to-factor => ]
    c-to-factor
(U) [ [ (get-catchstack) push ] dip call => (get-catchstack) pop* ]
(O) command-line-startup
(O) do-all
(O) do-tests
(O) test-vocabs
(O) test-vocab
(O) run-test-file
(U) [
        2 load-locals -1 get-local <experiment> load-local
        silent-tests? get 0 get-local
        [ load-local 0 get-local experiment. 1 drop-locals ] curry
        unless -2 get-local execute => 0 get-local -1 get-local [
            2 load-locals -1 get-local 0 get-local ~quotation~ curry
            curry ~quotation~ if 2 drop-locals
        ] curry curry [ 2drop "Warning: test skipped!" print ] if
        3 drop-locals
    ]
    experiment
(O) (unit-test)
(O) assert-sequence=
(O) assert-sequence
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

on linux/x86.64:

```
==== resource:extra/math/cardinality/cardinality-tests.factor

resource:extra/math/cardinality/cardinality-tests.factor: 14

Unit Test: {
    { t }
    [
        10 [
            10000
            [ random-units 10 estimate-cardinality ] [ / ] bi
        ] replicate [ 1.0 0.15 ~ ] all?
    ]
}

=== Expected:
t
=== Got:
f
(U) [ c-to-factor => ]
    c-to-factor
(U) [ [ (get-catchstack) push ] dip call => (get-catchstack) pop* ]
(O) command-line-startup
(O) do-all
(O) do-tests
(O) test-vocabs
(O) test-vocab
(O) run-test-file
(U) [
        2 load-locals -1 get-local <experiment> load-local
        silent-tests? get 0 get-local
        [ load-local 0 get-local experiment. 1 drop-locals ] curry
        unless -2 get-local execute => 0 get-local -1 get-local [
            2 load-locals -1 get-local 0 get-local ~quotation~ curry
            curry ~quotation~ if 2 drop-locals
        ] curry curry [ 2drop "Warning: test skipped!" print ] if
        3 drop-locals
    ]
    experiment
(O) (unit-test)
(O) assert-sequence=
(O) assert-sequence
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

on windows/x86.64:

```
==== resource:basis/tools/profiler/sampling/sampling-tests.factor

resource:basis/tools/profiler/sampling/sampling-tests.factor: 54

Unit Test: {
    { t t }
    [
        [ "resource:basis/tools/memory/memory.factor" run-file ]
        profile get-samples length 0 >
        OBJ-SAMPLE-CALLSTACKS special-object first 0 >
    ]
}

No suitable arithmetic method
left    1
right   f
generic +
(U) [ c-to-factor => ]
    c-to-factor
(U) [ [ (get-catchstack) push ] dip call => (get-catchstack) pop* ]
(O) command-line-startup
(O) do-all
(O) do-tests
(O) test-vocabs
(O) test-vocab
(O) run-test-file
(U) [
        { 5 } [ ~array~ 1 2 collect-tops keys first ] \ (unit-test)
        11 experiment
        { } [ 10 ~quotation~ times ] \ (unit-test) 15 experiment
        [ 10 ~quotation~ times ] [ boom? ] \ (must-fail-with) 16
        experiment
        { t t t t t t t t t t } [ 10 ~quotation~ times ]
        \ (unit-test) 24 experiment
        { t t t t t t t t t t } [ 10 ~quotation~ times ]
        \ (unit-test) 32 experiment
        { t t } [ 2 ~quotation~ times ] \ (unit-test) 39 experiment
        { t }
        [ ~quotation~ profile raw-profile-data get-global >boolean ]
        \ (unit-test) 44 experiment f raw-profile-data set-global gc
        { t t } [
            ~quotation~ profile get-samples length 0 >
            OBJ-SAMPLE-CALLSTACKS special-object first 0 >
        ] \ (unit-test) 54 experiment =>
        { t } [ gc ~quotation~ profile get-samples length 1 <= ]
        \ (unit-test) 61 experiment
    ]
(U) [
        2 load-locals -1 get-local <experiment> load-local
        silent-tests? get 0 get-local
        [ load-local 0 get-local experiment. 1 drop-locals ] curry
        unless -2 get-local execute => 0 get-local -1 get-local [
            2 load-locals -1 get-local 0 get-local ~quotation~ curry
            curry ~quotation~ if 2 drop-locals
        ] curry curry [ 2drop "Warning: test skipped!" print ] if
        3 drop-locals
    ]
    experiment
(O) (unit-test)
(U) [
        [ ~quotation~ dip swap ~quotation~ dip ] dip swap
        [ call get-datastack ] dip => swap [ set-datastack ] dip
    ]
(U) [ call => get-datastack ]
(U) [
        [ "resource:basis/tools/memory/memory.factor" run-file ]
        profile => get-samples length 0 >
        OBJ-SAMPLE-CALLSTACKS special-object first 0 >
    ]
(U) [ [ compose ~quotation~ curry recover ] [ drop ] 2bi => call ]
    cleanup
(U) [ [ 2keep ] dip => call ]
    2bi
(U) [ swap [ dip ] dip => ]
    2dip
(U) [ swap [ call ] dip => ]
    dip
(U) [ [ (get-catchstack) push ] dip call => (get-catchstack) pop* ]
(O) run-file
(O) parse-file
(O) parse-stream
(O) finish-compilation-unit
(O) M\ optimizing-compiler recompile
(O) compile-word
(O) frontend
(O) optimize-tree
(O) escape-analysis
(O) (escape-analysis)
(O) M\ #recursive escape-analysis*
(O) (escape-analysis)
(O) M\ #branch escape-analysis*
(O) (escape-analysis)
(O) M\ #branch escape-analysis*
(O) (escape-analysis)
(O) copy-values
(O) copy-value
(O) M\ disjoint-set equate
(O) +
(O) no-math-method
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

**Comment by erg** [comment]
trigger the bug:

```factor
10 [ [ "io.monitors" test ] in-thread ] times
```

**Comment by erg** [comment]
one bug per issue please 😆

**Comment by erg** [comment]
This one makes it fail. should be .85 to pass but it's .84

```factor
{ f } [
    4 <mersenne-twister> [
        3776 [ 10000 random-units drop ] times ! find the failing random numbers
        10000 [ random-units 10 estimate-cardinality ] [ / ] bi
        1.0 0.15 ~
    ] with-random
] unit-test
```

**Comment by erg** [comment]
fixed! open new issues for new bugs

**Comment by erg** [comment]
ONE MORE BUG

**Comment by mrjbq7** [comment]
Closing since 2/3 are fixed, opening the windows issue as #2747.
